### ISSUE TITLE
Pretrained models requested

### ISSUE THREAD

**Issue by sunset1995** [yhenon/pytorch-retinanet#2]
Hello,
Thanks for your simple and easy to modified code.
I'm currently working on a vision guidance system which built upon on this repo.
The result of trained from scratch from our dataset is acceptable to us but we want to further compared the result to fine-tuned from pretrained coco.

I have tried manually adapt the weight from [keras-retinanet](https://github.com/fizyr/keras-retinanet) (also the input scale and RGB to BGR preprocessing) but the bbox position is not good (see this [demo dog](https://imgur.com/WecbyOQ)).
It is wonderful if you can release your pretrained coco model.

Thanks you very much.

**Comment by JY-D** [comment]
I have the same issue here.
Looking for coco pre-trained model too.
Thanks.

**Comment by yhenon** [comment]
Hi,
I'm glad to hear you find this repo useful. I haven't shared weights yet as it's not quite finalized. However, I do have weights that I should be able to share later today. 
@sunset1995  the regression targets are normalized slightly differently in this repo VS the keras one. This one uses (0.1,0.1,0.2,0.2), whereas the keras one uses (0.2,0.2,0.2,0.2). So the weights for the final layer of the regression branch have to be scaled appropriately.

**Comment by sunset1995** [comment]
@yhenon thanks~
I also found the keras version output regression output is to shift `x1, y1, x2, y2` instead of `x, y, w, h`. The visualization from few samples look reasonable now.
I will more prefer to use your weight to keep code cleaner.

Thanks again~

**Comment by yhenon-nextdroid** [comment]
@sunset1995 a keras-retinanet -> pytorch-retinanet would be a very nice thing to have in this repo. Any chance you could submit a PR?

**Comment by yhenon** [comment]
@sunset1995 a keras-retinanet -> pytorch-retinanet would be a very nice
thing to have in this repo. Any chance you could submit a PR?

**Comment by sunset1995** [comment]
@yhenon I will open a new branch for it in this weekend.
Also, our team is waiting your pretrained weight :)

**Comment by yhenon** [comment]
I have uploaded a model here:
https://drive.google.com/open?id=1hCtM35R_t6T8RJVSd74K4gB-A1MR-TxC (it
reaches 33.5% mAP, just under the 34% mAP reported in the paper, but it is
trained with adam vs sgd with weight decay).

I'd appreciate if you could try it out before I add it to the README. You
should be able to `torch.load(filepath)` the model.

**Comment by sunset1995** [comment]
The results from few samples look prefect! (better than my keras version, so some fix is needed for my keras conversed model)
btw, it seem that `coco_resnet_50_map_0_335.pt` is saved from python2
I have to use below trick to load it from python3
```
from functools import partial
import pickle
pickle.load = partial(pickle.load, encoding="latin1")
pickle.Unpickler = partial(pickle.Unpickler, encoding="latin1")
retinanet = torch.load(parser.model, map_location=lambda storage, loc: storage, pickle_module=pickle)
```

**Comment by lvaleriu** [comment]
I see this line: 
            anchors_nms_idx = nms(torch.cat([transformed_anchors, scores], dim=2)[0, :, :], 0.5)
) that should apply nms to prediction detections too, but i get results like this: 
![image](https://user-images.githubusercontent.com/927999/42313934-2fa6e86a-8044-11e8-9d5c-7f49ca381cb2.png).

Do you have any idea why this is happening?

**Comment by yhenon** [comment]
I've uploaded a model that has been serialized differently (https://drive.google.com/open?id=1yLmjq3JtXi841yXWBxst0coAgR26MNBS).

You should be able to load it with:

```
retinanet = model.resnet50(num_classes=dataset_train.num_classes(),)
retinanet.load_state_dict(torch.load(PATH))
```

@sunset1995  This should fix the issues between python2/3

@lvaleriu this may fix the problem you reported, I'd appreciate if you could test it out.

**Comment by lvaleriu** [comment]
Same thing for me.. I get multiple anchors/detections for the same object

**Comment by wsnedy** [comment]
@yhenon hi, yhenon, could you explain a little bit that why you use (0.1,0.1,0.2,0.2) to normalize the regression targets but not use the original regression targets? It's because the regression loss is too small to be fully optimized? But in the original paper, it seems that the author doesn't normalize the regression targets, so I don't know if it's necessary. 
Thanks.

**Comment by hgaiser** [comment]
What do you mean with original regression targets ? If you compute the COCO regression mean and std values you will more or less get the 0.1 0.1 0.2 0.2 that is used. The original paper doesnt mention regression normalization but it is safe to assume they do. Also, as far as I know, Detectron (which is the closest to an official implementation) uses these regression values as well.

**Comment by wsnedy** [comment]
@hgaiser for the original regression targets, i mean the targets in [this line](https://github.com/yhenon/pytorch-retinanet/blob/master/losses.py#L119). And the purpose of doing regression normalization is to amplify the regression loss and to make it can be fully optimized, do i understand it correctly?

**Comment by yhenon** [comment]
@hgaiser is correct, I'll just add a few things.

- the original source for doing this is the fast-rcnn paper, where they
state:

```
We normalize the ground-truth regression targets vi to have zero mean and
unit variance. All experiments use λ = 1.
```

- these targets were computed over the whole COCO dataset.

>the purpose of doing regression normalization is to amplify the regression
loss and to make it can be fully optimized, do i understand it correctly?

partly, it also ensures that the 2 parts of the regression loss is balanced
(the size and position)

- the main reason to do it is that it improves mAP a little.

**Comment by yhenon** [comment]
@hgaiser is correct, I'll just add a few things.

- the original source for doing this is the fast-rcnn paper, where they
state:


>  We normalize the ground-truth regression targets vi to have zero mean and unit variance. All experiments use λ = 1.


- these targets were computed over the whole COCO dataset.

>the purpose of doing regression normalization is to amplify the regression
loss and to make it can be fully optimized, do i understand it correctly?

- partly, it also ensures that the 2 parts of the regression loss is balanced
(the size and position)

- the main reason to do it is that it improves mAP a little.

**Comment by wsnedy** [comment]
@yhenon it's clear to me now, thanks for your explanation.

**Comment by ObadaJabassini** [comment]
How many classes are supported in the pretrained model?

**Comment by xiaodiu2010** [comment]
@yhenon Thanks for your project. Great work!  
I wonder why SGD doesn't work as the original paper said?  Adam works well.
Do you have any ideas?
